src/metabase/db/custom_migrations.clj

```diff
@@ -52,3 +58,49 @@
   "Define a custom migration."
   [name migration-body]
   `(def-reversible-migration ~name ~migration-body ([~'_] (no-op ~(str name)))))
+
+
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                                  MIGRATIONS                                                    |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(def ^:private base-path-regex
+  #"^(/db/\d+(?:/schema/(?:(?:[^\\/])|(?:\\/)|(?:\\\\))*(?:/table/\d+?)?)?/)((native/)|(query/(segmented/)?))?$")
+
+(defn- ->v2-paths
+  "Converts v1 data permission paths into v2 data and query permissions paths. This is similar to `->v2-path` in
+   metabase.models.permissions but somewhat simplified for the migration use case."
+  [v1-path]
+  (if-let [base-path (second (re-find base-path-regex v1-path))]
+    ;; For (almost) all v1 data paths, we simply extract the base path (e.g. "/db/1/schema/PUBLIC/table/1/")
+    ;; and construct new v2 paths by adding prefixes to the base path.
+    [(str "/data" base-path) (str "/query" base-path)]
+
+    ;; For the specific v1 path that grants full data access but no native query access, we add a
+    ;; /schema/ suffix to the corresponding v2 query permission path.
+    (when-let [db-id (second (re-find #"^/db/(\d+)/schema/$" v1-path))]
+      [(str "/data/db/" db-id "/") (str "/query/db/" db-id "/schema/")])))
+
+(def-reversible-migration SplitDataPermissions
```

Comment:
Not really part of this PR but traditionally `def` macros that consist of more than one word spell out the whole word `define`, so this should have been called `define-reversible-migration`

Reply:
👍 I can change it in this PR